### PR TITLE
Adds the Hydra configuration system for RL training

### PR DESCRIPTION
**Pull request by Dhoeller19**
# Description

This MR adds utilities to enable the hydra configuration system.
Using the `train.py` scripts, the user can now change any parameter in the environment or agent configs from command line inputs, for example:
``` 
python source/standalone/workflows/rsl_rl/train.py --task Isaac-Cartpole-v0 --headless env.actions.joint_effort.scale=10.0
```

## Type of change

- New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I have run the [`pre-commit` checks](https://pre-commit.com/) with `./isaaclab.sh --format`
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have updated the changelog and the corresponding version in the extension's `config/extension.toml` file
- [ ] I have added my name to the `CONTRIBUTORS.md` or my name already exists there

### CODE REVIEW

File: source/extensions/omni.isaac.lab/omni/isaac/lab/utils/dict.py [hunk 1]
```diff
@@ -253,3 +286,34 @@ def print_dict(val, nesting: int = -4, start: bool = True):
...
         else:
             print(val)
+
+
+"""
+Helper functions.
+"""
+
+
+def string_to_slice(s):
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/utils/dict.py]: Makes mores sense to move this here?

https://github.com/isaac-sim/IsaacLab/blob/main/source/extensions/omni.isaac.lab/omni/isaac/lab/utils/string.py

File: source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py [hunk 2]
```diff
@@ -107,7 +107,7 @@ def reset(self, env_ids: Sequence[int] | None = None) -> dict[str, torch.Tensor]
             # store information
             # r_1 + r_2 + ... + r_n
             episodic_sum_avg = torch.mean(self._episode_sums[key][env_ids])
-            extras["Episode Reward/" + key] = episodic_sum_avg / self._env.max_episode_length_s
+            extras["Episode_Reward/" + key] = episodic_sum_avg / self._env.max_episode_length_s
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py:110]: Why do we even have "Episode" as a prefix. Shall we just call it Reward? Also this is breaking again since all the prior logs will appear differently :( 

Since this is not Hydra-related, can we split it in a different MR with a reason to have this change?

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+def hydra_task_config(task_name: str, agent_cfg_entry_point: str) -> Callable:
+    """Decorator to handle the Hydra configuration for a task.
+
+    This decorator registers the task to Hydra and updates the environment and agent configurations from Hydra parsed
+    command line arguments.
+
+    Args:
+        task_name (str): The name of the task.
+        agent_cfg_entry_point (str): The entry point key to resolve the agent's configuration file.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py]: We only do type-hinting in the signature (as it is easier to maintain).

```suggestion
        task_name: The name of the task.
        agent_cfg_entry_point: The entry point key to resolve the agent's configuration file.
```

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    task_name: str, agent_cfg_entry_point: str
+) -> tuple[ManagerBasedRLEnvCfg | DirectRLEnvCfg, dict]:
+    """Register the task configuration to the Hydra configuration store.
+
+    This function resolves the configuration file for the environment and agent based on the task's name.
+    It then registers the configurations to the Hydra configuration store.
+
+    Args:
+        task_name (str): The name of the task.
+        agent_cfg_entry_point (str): The entry point key to resolve the agent's configuration file.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py]: We only do type-hinting in the signature (as it is easier to maintain).

```suggestion
        task_name: The name of the task.
        agent_cfg_entry_point: The entry point key to resolve the agent's configuration file.
```

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+    Returns:
+        The parsed environment and agent configurations.
```
Comment (Mayankm96) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py]:
```suggestion
        A tuple containing the parsed environment and agent configuration objects.
```

File: source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py (hunk 2 above)
Comment (Dhoeller19) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py:110]: This is somehow hydra related, I get all of a sudden some weird tensorboard warnings that "Episode Reward" is not a valid string.

Comment (Mayankm96) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py:110]: Huh.... Why does hydra care about tensorboard :/ . I won't have time this Thursday to check it out but I feel this shouldn't be affected by Hydra o.O

Comment (Dhoeller19) [reply source/extensions/omni.isaac.lab/omni/isaac/lab/managers/reward_manager.py:110]: yes this is super strange

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    env_cfg_dict = env_cfg.to_dict()
+    if isinstance(agent_cfg, dict):
+        agent_cfg_dict = agent_cfg
+    else:
+        agent_cfg_dict = agent_cfg.to_dict()
+    cfg_dict = {"env": env_cfg_dict, "agent": agent_cfg_dict}
+    # replace slices with strings because OmegaConf does not support slices
+    cfg_dict = replace_slices_with_strings(cfg_dict)
+    # store the configuration to Hydra
+    ConfigStore.instance().store(name=task_name, node=cfg_dict)
```
Comment (romesco) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py:53]: This might be intended, but just as a heads up, if we pass a dict (rather than a structured config), it forgoes the type verification that hydra typically performs: https://hydra.cc/docs/tutorials/structured_config/config_store/

File: source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    """
+
+    def decorator(func):
+        @functools.wraps(func)
+        def wrapper(*args, **kwargs):
+            # register the task to Hydra
+            env_cfg, agent_cfg = register_task_to_hydra(task_name, agent_cfg_entry_point)
+
+            # define thr new Hydra main function
+            @hydra.main(config_path=None, config_name=task_name, version_base="1.3")
```
Comment (romesco) [inline source/extensions/omni.isaac.lab_tasks/omni/isaac/lab_tasks/utils/hydra.py:78]: Here, it is also possible to use the compose API: https://hydra.cc/docs/advanced/compose_api/

The minimal example originally done for ray: https://github.com/facebookresearch/hydra/blob/main/examples/advanced/ray_example/ray_compose_example.py

File: docs/source/features/hydra.rst [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+.. note::
+
+    To keep backwards compatibility, and to provide a more user-friendly experience, we have kept the old cli arguments
+    of the form ``--param``, for example ``--num_envs``, ``--seed``, ``--max_iterations``. These arguments have precedence
+    over the hydra arguments, and will overwrite the values set by the hydra arguments.
+
+.. attention::
+
+    Particular care should be taken when modifying the parameters using command line arguments. Some of the configurations
```
Comment (romesco) [inline docs/source/features/hydra.rst:58]: This can be addressed using OmegaConf resolvers if we want to provide a "hydra-friendly" patch to the configs. I can follow up here with a simple example.

**Review by romesco — COMMENTED** [review]
Some high-level comments, but I will follow up with some snippets.

Also general note: Currently, this PR leaves a lot on the table w.r.t the benefits of typing and config groups, but I understand that the minimum viable integration is important to get shipped first!

**Review by kellyguo11 — APPROVED** [review]

**Comment by kellyguo11** [comment]
Thanks for the review and recommendations @romesco! I'll get this one merged in first as the initial integration, please feel free to create follow up pull requests.
